File: November 2022/12 November 2022/Python/Maximum Sub-String after at most K changes.py
#User function Template for python3

class Solution:
	def characterReplacement(self, s, k):
        # Code here
        has={}
        maxi=0
        let=""
        left=0
        right=0
        res=0
        while right<len(s):
            if s[right] not in has:
                has[s[right]]=1
            else:
                has[s[right]]+=1
            maxi=max(maxi,has[s[right]])
            if right-left+1-maxi>k:
                has[s[left]]-=1
                left+=1
            else:
                res=max(res,right-left+1)
            right+=1
        return res

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		s = input()
		k = int(input())
		ob = Solution()
		ans = ob.characterReplacement(s, k)
		print(ans)

# } Driver Code Ends

File: Problems & Solutions/M/Maximum Sub-String after at most K changes/12 November 2022/Java/Maximum Sub-String after at most K changes.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            String s = br.readLine().trim();
            int k = Integer.parseInt(br.readLine().trim());
            Solution ob = new Solution();
            int ans = ob.characterReplacement(s, k);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution
{
    public int characterReplacement(String s, int k)
    {
        int start=0,max=0,ans=0;
        int []arr=new int[26];
        for(int i=0;i<s.length();i++){
            arr[s.charAt(i)-'A']+=1;
            max=Math.max(max,arr[s.charAt(i)-'A']);
            if(Math.abs(max-(i-start+1))>k){
                arr[s.charAt(start)-'A']-=1;
                max=0;
                start++;
                for(int ele:arr) max=Math.max(ele,max);
            }
            ans=Math.max(ans,i-start+1);
        }
        return ans;
    }
}